File: 215.kth-largest-element-in-an-array.go
/*
 * @lc app=leetcode id=215 lang=golang
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h IntHeap) Top() int {
	return h[0]
}
func (h *IntHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func findKthLargest(nums []int, k int) int {
	h := &IntHeap{}
	heap.Init(h)
	for _, num := range nums {
		if h.Len() < k {
			heap.Push(h, num)
		} else if (*h)[0] < num {
			heap.Pop(h)
			heap.Push(h, num)
		}
	}
	return heap.Pop(h).(int)
}
// @lc code=end

File: 215.kth-largest-element-in-an-array.js
/*
 * @lc app=leetcode id=215 lang=javascript
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function (nums, k) {
    // solution 1
    // O(nlogk)
    const pq = new MinPriorityQueue();
    for (const num of nums) {
        pq.enqueue(num);
        if (pq.size() > k) {
            pq.dequeue();
        }
    }
    return pq.dequeue().element;

    // // solution 2
    // // O(nlogn)
    // nums.sort((a, b) => a - b);
    // return nums[nums.length - k];
};
// @lc code=end

